xdsl/dialects/csl.py

```diff
@@ -80,6 +83,84 @@ class ModuleKind(StrEnum):
     PROGRAM = "program"
 
 
+class TaskKind(StrEnum):
+    LOCAL = "local"
+    DATA = "data"
+    CONTROL = "control"
+
+
+class _FuncBase:
+    """
+    Base class for the shared functionalty of FuncOp and TaskOp
+    """
+
+    body: Region = region_def()
+    sym_name: StringAttr = prop_def(StringAttr)
+    function_type: FunctionType = prop_def(FunctionType)
+    arg_attrs = opt_prop_def(ArrayAttr[DictionaryAttr])
+    res_attrs = opt_prop_def(ArrayAttr[DictionaryAttr])
+
+    @abstractmethod
+    def _get_name(self) -> str: ...
+    @abstractmethod
+    def _get_attribs_and_props(self) -> dict[str, Attribute]: ...
```

Comment:
Another minor nit comment:

I don't quite understand why these are needed here, as they are both implemented as just getters for the `name` attribute, which you can just grab the same way using `self.name`

Also, I would make `_FuncBase` actually inherit from `ABC` *and* `IRDLOperation` so that it is a proper abstract class and can access all the IRDL fields (such as name).

Reply:
Now that I am looking at it, I am not entirely sure why _FuncBase doesn't just inherit from `IRDLOperation`. I think I ran into some issues early on when I didn't quite get how the `@irdl_op_definition` worked.